### PR TITLE
Switch to GAP 4.12dev

### PR DESCRIPTION
**Pull request by fingolfin**

### CODE REVIEW

**Comment by fingolfin** [comment]
Fun, lots of different errors:
- with Julia <= 1.5 on macOS: fails to load because Readline_jll 8.1.1 is only available in Julia >= 1.6; fix will come via https://github.com/JuliaPackaging/Yggdrasil/pull/3217
- with Julia <= 1.5 on macOS: "Error, Matrix Element: <mat>[2,1] must have an assigned value" in `test/conversion.jl:143` -- I think we changed behaviour in GAP and just need to disable that test
- with Linux: `LoadError: InitError: could not load symbol "julia_gap"` -- this may be caused by https://github.com/gap-system/gap/pull/4234

File: pkg/JuliaInterface/src/JuliaInterface.c [hunk 1]
```diff
@@ -319,22 +319,21 @@ static Obj IsOutputStream;
...
-    syJmp_buf readJmpError;
-
     BEGIN_GAP_SYNC();
 
     if (CALL_1ARGS(IsOutputStream, stream) != True) {
         ErrorQuit("STREAM_CALL: <outstream> must be an output stream", 0, 0);
     }
-    if (!OpenOutputStream(stream)) {
+    TypOutputFile output = { 0 };
+    if (!OpenOutputStream(&output, stream)) {
```
Comment (fingolfin) [inline pkg/JuliaInterface/src/JuliaInterface.c:328]: This is really brittle; we should either move `STREAM_CALL` to GAP, or else add a libgap-api that does the brunt of its work for us

**Comment by fingolfin** [comment]
Not sure where the "Cannot write to file descriptor 0" errors are coming from now. I hope https://github.com/gap-system/gap/pull/4589 will help with that. Once it and https://github.com/gap-system/gap/pull/4587 land, I'll update GAP_jll once more and continue with investigating this.

**Comment by fingolfin** [comment]
Hrm, I've updated to a slightly newer GAP_jll now, with https://github.com/gap-system/gap/pull/4600 in it -- and now it got through the Julia tests! But now it crashes (with no helpful backtrace) in the "GAP tests" when building the manual, it seems:
```
#I Producing the index . . .
#I Second run through document . . .
#I  File: ./doc/manual.lab written.
Extracting manual examples for JuliaInterface package ...
3 chapters detected
Chapter 1 : no examples 
Chapter 2 : extracted 19 examples
Chapter 3 : extracted 4 examples
etc/ci_test.sh: line 12:  6875 Segmentation fault: 11  ${GAP} makedoc.g
Error: Process completed with exit code 139.
```

**Comment by ThomasBreuer** [comment]
I can reproduce the segmentation fault on my notebook (ubuntu 20.04.1 + Julia 1.5.3) when I call
`[...]/GAP.jl/gaproot/v1.5/bin/gap.sh -A --quitonbreak --norepl makedoc.g`.
The problem does *not* occur when I start Julia, enter `using GAP; GAP.prompt()`, and then call `Read("makedoc.g");`.

**Comment by fingolfin** [comment]
Yeah I can reproduce it, too. But now I need to compile a debug version of Julia to be able to attach a debugger *sigh*

**Comment by fingolfin** [comment]
Urhg, trying with `julia-debug`, precompiling GAP fails. Even on master. In fact, just loading GAP_jll triggers a crash:
```
julia> using GAP_jll
Two passes with the same argument (-LowerPTLS) attempted to be registered!

signal (4): Illegal instruction: 4
in expression starting at REPL[4]:1
_ZN4llvm14PassNameParser14passRegisteredEPKNS_8PassInfoE at /Users/mhorn/Projekte/Julia/julia.release-1.6/usr/lib/libLLVM.dylib (unknown line)
_ZN4llvm12PassRegistry12registerPassERKNS_8PassInfoEb at /Users/mhorn/Projekte/Julia/julia.release-1.6/usr/lib/libLLVM.dylib (unknown line)
RegisterPass at /Users/mhorn/Projekte/Julia/julia.release-1.6/usr/include/llvm/PassSupport.h:101 [inlined]
RegisterPass at /Users/mhorn/Projekte/Julia/julia.release-1.6/usr/include/llvm/PassSupport.h:100 [inlined]
__cxx_global_var_init at /Users/mhorn/Projekte/Julia/julia.release-1.6/src/llvm-ptls.cpp:273 [inlined]
unknown function (ip: 0x1570a436d)
Allocations: 8796281 (Pool: 8792559; Big: 3722); GC: 12
Illegal instruction: 4
```

Most likely this comes from the init section in GAP_jll, which does this:
```gap
    sym = dlsym(libgap_handle, :GAP_InitJuliaMemoryInterface)
    ccall(sym, Nothing, (Any, Ptr{Nothing}), @__MODULE__, C_NULL)
```


The macOS crash reported actually gives a bit longer backtrace (also with C++ unmangling applied):
```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libLLVM.dylib                 	0x000000010a399277 llvm::PassNameParser::passRegistered(llvm::PassInfo const*) + 359
1   libLLVM.dylib                 	0x000000010a3a26ec llvm::PassRegistry::registerPass(llvm::PassInfo const&, bool) + 268
2   libjulia-internal.1.dylib     	0x00000001556af96e _GLOBAL__sub_I_llvm_ptls.cpp + 158
3   dyld                          	0x000000010f887592 ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 506
4   dyld                          	0x000000010f887798 ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 40
5   dyld                          	0x000000010f882bea ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 362
6   dyld                          	0x000000010f881d73 ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 133
7   dyld                          	0x000000010f881e05 ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 73
8   dyld                          	0x000000010f874cb2 dyld::runInitializers(ImageLoader*) + 82
9   dyld                          	0x000000010f87e3dc dlopen_internal + 607
10  libdyld.dylib                 	0x00007fff63cfad43 dlopen + 200
11  libjulia.dylib                	0x000000010f73e7c7 load_library + 135 (loader_lib.c:51)
12  libjulia.dylib                	0x000000010f73e6df jl_load_libjulia_internal + 127 (loader_lib.c:159)
13  dyld                          	0x000000010f887592 ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 506
14  dyld                          	0x000000010f887798 ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 40
15  dyld                          	0x000000010f882bea ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 362
16  dyld                          	0x000000010f882b80 ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 256
17  dyld                          	0x000000010f881d73 ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 133
18  dyld                          	0x000000010f881e05 ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 73
19  dyld                          	0x000000010f874cb2 dyld::runInitializers(ImageLoader*) + 82
20  dyld                          	0x000000010f87e3dc dlopen_internal + 607
21  libdyld.dylib                 	0x00007fff63cfad43 dlopen + 200
22  libjulia-internal-debug.1.dylib	0x00000001093ab241 jl_dlopen + 177 (dlload.c:124)
23  libjulia-internal-debug.1.dylib	0x00000001093ab7ab jl_load_dynamic_library + 1291 (dlload.c:257)
24  sys-debug.dylib               	0x000000010fb31fec julia_#dlopen#3_26333 + 188 (libdl.jl:115)
25  ???                           	0x000000010f3447fa 0 + 4550051834
26  ???                           	0x000000010f344910 0 + 4550052112
27  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
28  ???                           	0x000000010f345266 0 + 4550054502
29  libjulia-internal-debug.1.dylib	0x000000010937fae7 jl_fptr_args + 71 (gf.c:2006)
30  libjulia-internal-debug.1.dylib	0x0000000109389564 _jl_invoke + 388 (gf.c:2237)
31  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
32  libjulia-internal-debug.1.dylib	0x00000001093cf5f2 jl_apply + 50 (julia.h:1703)
33  libjulia-internal-debug.1.dylib	0x00000001093cf4a8 jl_module_run_initializer + 168 (toplevel.c:73)
34  libjulia-internal-debug.1.dylib	0x00000001093b8061 jl_init_restored_modules + 129 (dump.c:2110)
35  sys-debug.dylib               	0x000000010fa500ae japi1__include_from_serialized_44085 + 510
36  sys-debug.dylib               	0x000000010fb7eacd julia__require_search_from_serialized_44547 + 5453 (loading.jl:760)
37  sys-debug.dylib               	0x000000010fd05161 julia__require_49035 + 1361 (loading.jl:999)
38  sys-debug.dylib               	0x000000010fbf40c0 julia_require_33613 + 128 (loading.jl:914)
39  sys-debug.dylib               	0x000000010fbf51d3 japi1_require_33619 + 3747 (loading.jl:901)
40  libjulia-internal-debug.1.dylib	0x000000010937fae7 jl_fptr_args + 71 (gf.c:2006)
41  libjulia-internal-debug.1.dylib	0x0000000109389475 _jl_invoke + 149 (gf.c:2218)
42  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
43  libjulia-internal-debug.1.dylib	0x00000001093cf5f2 jl_apply + 50 (julia.h:1703)
44  libjulia-internal-debug.1.dylib	0x00000001093d4297 call_require + 263 (toplevel.c:421)
45  libjulia-internal-debug.1.dylib	0x00000001093d261d eval_import_path + 349 (toplevel.c:458)
46  libjulia-internal-debug.1.dylib	0x00000001093d04a0 jl_toplevel_eval_flex + 1408 (toplevel.c:684)
47  libjulia-internal-debug.1.dylib	0x00000001093d1f81 jl_eval_module_expr + 1393 (toplevel.c:195)
48  libjulia-internal-debug.1.dylib	0x00000001093d0342 jl_toplevel_eval_flex + 1058 (toplevel.c:668)
49  libjulia-internal-debug.1.dylib	0x00000001093d0fbc jl_toplevel_eval_flex + 4252 (toplevel.c:825)
50  libjulia-internal-debug.1.dylib	0x00000001093d0fbc jl_toplevel_eval_flex + 4252 (toplevel.c:825)
51  libjulia-internal-debug.1.dylib	0x00000001093d2c04 jl_toplevel_eval + 36 (toplevel.c:886)
52  libjulia-internal-debug.1.dylib	0x00000001093d2cda jl_toplevel_eval_in + 202 (toplevel.c:929)
53  sys-debug.dylib               	0x000000010f9d6be8 japi1_include_string_37379 + 728 (loading.jl:1094)
54  libjulia-internal-debug.1.dylib	0x000000010937fae7 jl_fptr_args + 71 (gf.c:2006)
55  libjulia-internal-debug.1.dylib	0x0000000109389475 _jl_invoke + 149 (gf.c:2218)
56  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
57  sys-debug.dylib               	0x000000010fc30123 japi1__include_43013 + 1411 (loading.jl:1148)
58  sys-debug.dylib               	0x000000010fc57227 julia_include_package_for_output_41999 + 2535 (loading.jl:1213)
59  sys-debug.dylib               	0x000000010fc57646 jfptr_include_package_for_output_42000 + 38
60  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
61  libjulia-internal-debug.1.dylib	0x00000001093aa6d2 jl_apply + 50 (julia.h:1703)
62  libjulia-internal-debug.1.dylib	0x00000001093aa348 do_call + 312 (interpreter.c:115)
63  libjulia-internal-debug.1.dylib	0x00000001093a8c85 eval_value + 1205 (interpreter.c:204)
64  libjulia-internal-debug.1.dylib	0x00000001093a9c6d eval_stmt_value + 29 (interpreter.c:155)
65  libjulia-internal-debug.1.dylib	0x00000001093a8122 eval_body + 3986
66  libjulia-internal-debug.1.dylib	0x00000001093a8572 jl_interpret_toplevel_thunk + 450 (interpreter.c:670)
67  libjulia-internal-debug.1.dylib	0x00000001093d14bd jl_toplevel_eval_flex + 5533 (toplevel.c:877)
68  libjulia-internal-debug.1.dylib	0x00000001093d2c04 jl_toplevel_eval + 36 (toplevel.c:886)
69  libjulia-internal-debug.1.dylib	0x00000001093d2cda jl_toplevel_eval_in + 202 (toplevel.c:929)
70  sys-debug.dylib               	0x000000010ffe76d3 japi1_eval_34401 + 35 (client.jl:446)
71  libjulia-internal-debug.1.dylib	0x0000000109389475 _jl_invoke + 149 (gf.c:2218)
72  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
73  libjulia-internal-debug.1.dylib	0x00000001093aa6d2 jl_apply + 50 (julia.h:1703)
74  libjulia-internal-debug.1.dylib	0x00000001093aa348 do_call + 312 (interpreter.c:115)
75  libjulia-internal-debug.1.dylib	0x00000001093a8c85 eval_value + 1205 (interpreter.c:204)
76  libjulia-internal-debug.1.dylib	0x00000001093a9c6d eval_stmt_value + 29 (interpreter.c:155)
77  libjulia-internal-debug.1.dylib	0x00000001093a8122 eval_body + 3986
78  libjulia-internal-debug.1.dylib	0x00000001093a8572 jl_interpret_toplevel_thunk + 450 (interpreter.c:670)
79  libjulia-internal-debug.1.dylib	0x00000001093d14bd jl_toplevel_eval_flex + 5533 (toplevel.c:877)
80  libjulia-internal-debug.1.dylib	0x00000001093d0fbc jl_toplevel_eval_flex + 4252 (toplevel.c:825)
81  libjulia-internal-debug.1.dylib	0x00000001093d2c04 jl_toplevel_eval + 36 (toplevel.c:886)
82  libjulia-internal-debug.1.dylib	0x00000001093d2cda jl_toplevel_eval_in + 202 (toplevel.c:929)
83  sys-debug.dylib               	0x000000010fea713d julia_exec_options_34051 + 7197 (client.jl:261)
84  sys-debug.dylib               	0x000000010fea902d julia__start_23799 + 685 (client.jl:485)
85  libjulia-internal-debug.1.dylib	0x0000000109389605 jl_apply_generic + 117 (gf.c:2419)
86  libjulia-internal-debug.1.dylib	0x0000000109402782 jl_apply + 50 (julia.h:1703)
87  libjulia-internal-debug.1.dylib	0x0000000109403a88 true_main + 280 (jlapi.c:561)
88  libjulia-internal-debug.1.dylib	0x0000000109403933 repl_entrypoint + 371 (jlapi.c:702)
89  libjulia-debug.1.6.dylib      	0x00000001091d68a0 jl_load_repl + 144 (loader_lib.c:199)
90  julia-debug                   	0x00000001091cef7b main + 27 (loader_exe.c:51)
91  libdyld.dylib                 	0x00007fff63d0f3d5 start + 1
```

**Comment by fingolfin** [comment]
Ahh, but after updating the Julia branch `release-1.6` to the latest (i.e., the code which will soon become Julia 1.6.2) this crash seems to be gone. Puh, so it's perhaps not our fault, good :-)

**Comment by fingolfin** [comment]
Note that normally, GAP finishes after printing `Chapter 3 : extracted 4 examples`. So I now suspect this is something in the cleanup stage...

When it crashes, there is no useful backtrace; but it seems to be doing a `longjmp`, which means `GAP_THROW` is suspect...
```
Process 11085 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00007fff63efab20 libsystem_platform.dylib`_longjmp + 84
libsystem_platform.dylib`_longjmp:
->  0x7fff63efab20 <+84>: jmpq   *%rsi

libsystem_platform.dylib`_sigunaltstack:
    0x7fff63efab22 <+0>:  cmpl   $0x1, %edi
    0x7fff63efab25 <+3>:  movl   $0x40000000, %eax         ; imm = 0x40000000
    0x7fff63efab2a <+8>:  movl   $0x80000000, %esi         ; imm = 0x80000000
Target 0: (julia) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x00007fff63efab20 libsystem_platform.dylib`_longjmp + 84
```

Perhaps we are neglecting setup a `GAP_TRY` in a place where we call into GAP? Note that we replaced `STREAM_CALL` (which used `Call1ArgsInNewReader` and thus "swallowed" GAP exceptions) by `CALL_WITH_STREAM` (which doesn't do this).

So I set a breakpoint on `GAP_THROW`. It is called three times before the crash, each time from inside `EvalFunccall1args`. That one only ever calls it by this code in it:
```C
    if (STATE(UserHasQuit) || STATE(UserHasQUIT)) {
        // the function must have called READ() and the user quit from
        // a break loop inside it
        GAP_THROW();
    }
```
So that confirms my suspicion that this happens while GAP is exiting.

Now the first two times, the GAP part of the backtrace starts with `Call0ArgsInNewReader`; the third time via `DoExecFunc1args`. 

Interestingly, I have now verified that it dies at the very end, when trying to exit

**Comment by fingolfin** [comment]
.... and looking at the exit code, I now recall that I had to do some weird stuff because it was not possible to retrieve GAP's stored exit status code; but in GAP 4.12, this is actually possible now. So I changed the code to use that, and this fixes the issue -- at least locally for me. Let's see what CI says

**Comment by fingolfin** [comment]
@ThomasBreuer yay it works now! Except for failures in the JuliaExperimental test where I see diffs like this:
```patch
diff --git a/pkg/JuliaExperimental/tst/numfield.tst b/pkg/JuliaExperimental/tst/numfield.tst
index 8bc6c4a..a216c97 100644
--- a/pkg/JuliaExperimental/tst/numfield.tst
+++ b/pkg/JuliaExperimental/tst/numfield.tst
@@ -13,11 +13,11 @@ gap> z:= Zero( f );;  o:= One( f );;  a:= RootOfDefiningPolynomial( f );;
 gap> v:= [ z, o ];
 [ !0, !1 ]
 gap> nv:= GAPToNemo( c, v );
-<<Julia: [0 1]>>
+<vector object of length 2 over Monoid( ... )>
 gap> mat:= [ [ o, a/2 ], [ z, o ] ];
 [ [ !1, 1/2*a ], [ !0, !1 ] ]
 gap> nmat:= GAPToNemo( c, mat );
-<<Julia: [1 1//2*a; 0 1]>>
+<matrix object of dimensions 2x2 over Monoid( ... )>
 
 #gap> JuliaTypeInfo( JuliaPointer( nmat ) );
 #"AbstractAlgebra.Generic.Mat{Nemo.nf_elem}"
```
I will adapt the test for this for now, but perhaps it means we want to change something else there -- you'll know that better than me (or we can discuss, of course!)

File: pkg/JuliaExperimental/tst/numfield.tst [hunk 2]
```diff
@@ -13,50 +13,54 @@ gap> z:= Zero( f );;  o:= One( f );;  a:= RootOfDefiningPolynomial( f );;
...
 #gap> JuliaTypeInfo( JuliaPointer( nmat ) );
 #"AbstractAlgebra.Generic.Mat{Nemo.nf_elem}"
 gap> PrintObj( nmat );  Print( "\n" );
 [1 1//2*a; 0 1]
-gap> IsZero( nmat );
-false
-gap> IsZero( 0 * nmat );
-true
+gap> IsZero( nmat ); # TODO/FIXME
+Error, no method found! For debugging hints type ?Recovery from NoMethodFound
```
Comment (fingolfin) [inline pkg/JuliaExperimental/tst/numfield.tst:27]: I guess this means there is still work we should do on the GAP side on the MatrixObj code -- and/or we should add some methods for MatrixObj here?

In any case, I put all these changes into a separate commit so that you can revert it easily.

**Comment by fingolfin** [comment]
Still fails, because the GAP package `profiling` fails to compile... *sigh*

**Comment by fingolfin** [comment]
I've now resolved the JuliaExperimental issues, it all came down to ranking. My "fix" is of course not ideal, but it's good enough for this "experimental" code, I think :-)
